kubernetes: Document new common labels (#100)
giantswarm/fmt: kubernetes/annotations_and_labels.md

  managed by `chart-operator`.
- `release.giantswarm.io/version` - value should be Giant Swarm release
  version, e.g. `release.giantswarm.io/version=2.3.0`.
- `OPERATOR.giantswarm.io/version` - value should be the version of the
  operator as defined in `project.go` and matching `appVersion` in
  `Chart.yaml`, e.g. `kvm-operator.giantswarm.io/version=1.0.0`. It is used by
  the given operator to recognize which object it should reconcile (i.e. to

>   only reconcile objects matching its own version). When set on nodes it is

Reviewer: ```suggestion
  only reconcile objects matching its own version). When set on Node objects it is
```

Author: Done

  used to set that information in the status with the statusresource. This is
  different from release version and can be the same in multiple releases. Its
  value may be equal to that of `app.kubernetes.io/version` but it has a
  different purpose and since there could be multiple operators reconciling one
  object there could be multiple per-operator labels on one object.
- `app.giantswarm.io/branch` - (informational) branch from which this
  instance of the app/operator was built from.

> - `app.giantswarm.io/commit` - (informational) ID of the commit from which

Reviewer: ```suggestion
- `app.giantswarm.io/commit` - (informational) git sha of the commit from which
```

Author: Added it in brackets, I think SHA is just an implementation detail that [may hopefully go away one day](https://lwn.net/Articles/715716/). I believe git internally has mostly gone through refactoring and refers to those as just object IDs.

  managed by `chart-operator`.
- `release.giantswarm.io/version` - value should be Giant Swarm release
  version, e.g. `release.giantswarm.io/version=2.3.0`.
- `app.giantswarm.io/version` - value should be an app/operator version as
  defined in `project.go`, e.g. `app.giantswarm.io/version=1.0.0`. This is not
  the same as release version. This value can be the same in multiple releases.
  This label is used by the operator to recognize which object it should
  reconcile. When set on nodes it is used to set that information in the status

>   with the statusresource.

Reviewer: Maybe we should also add a word that this should match `appVersion` in Chart.yaml.

Author: Done.

  managed by `chart-operator`.
- `release.giantswarm.io/version` - value should be Giant Swarm release
  version, e.g. `release.giantswarm.io/version=2.3.0`.
- `OPERATOR.giantswarm.io/version` - value should be the version of the
  operator as defined in `project.go` and matching `appVersion` in
  `Chart.yaml`, e.g. `kvm-operator.giantswarm.io/version=1.0.0`. It is used by
  the given operator to recognize which object it should reconcile (i.e. to
  only reconcile objects matching its own version). When set on nodes it is
  used to set that information in the status with the statusresource. This is
  different from release version and can be the same in multiple releases. Its

>   value may be equal to that of `app.kubernetes.io/version` but it has a

Author: Actually scratch that, need to revert it as it is only equal if both labels refer to the same operator (i.e. the primary one), if there's a secondary operator reconciling an object they will be different.

  different purpose and since there could be multiple operators reconciling one
  object there could be multiple per-operator labels on one object.
- `app.kubernetes.io/name`, `app.kubernetes.io/version` and other
  [common labels][k8s-common-labels] - are informational, usually applied to
  all objects in a chart that installs an app or operator and so they indicate
  the source of that app/operator; value of `.../version` should be the
  app/operator version as defined in `project.go` and matching `appVersion` in

>   `Chart.yaml`, e.g. `app.kubernetes.io/version=1.0.0`.

Reviewer: Should we mention that here? We have `app` and `version` already. Do you think we should change that in scope of this task? Should we have both? Should we actively migrate to the official recommendation?

Author: I wanted to add them for a couple of reasons, mainly that `version` is not the same as `app.kubernetes.io/version` and the second thing was that we decided to drop `app.giantswarm.io/version` which I totally agree as it would just be duplicating the upstream labels, so I feel it's better to mention them here to avoid that. Also there was interest from people to adopt them anyway, but with understanding that they won't replace our internal labels at least not immediately.

Reviewer: Yes I think when we introduce something new we should also have a migration plan and actively do this. If you want to do this that's awesome but I'd say it's unrelated to this task.

Reviewer: Let's have a chat on google meet on Slack.

- `release.giantswarm.io/version` - value should be Giant Swarm release
  version, e.g. `release.giantswarm.io/version=2.3.0`.
- `app.giantswarm.io/version` - value should be an app/operator version as
  defined in `project.go`, e.g. `app.giantswarm.io/version=1.0.0`. This is not
  the same as release version. This value can be the same in multiple releases.
  This label is used by the operator to recognize which object it should
  reconcile. When set on nodes it is used to set that information in the status
  with the statusresource.
- `app.giantswarm.io/branch` - (informational) branch from which this
  instance of the app/operator was built from.
- `app.giantswarm.io/commit` - (informational) ID of the commit from which
  this instance of the app/operator was built from.
- `giantswarm.io/provider` - value should be the installation's provider, e.g.
  `kvm`, `aws`, or `azure`.
- `version` - value should contain the version of the application.  Should be applied
  to every Kubernetes resource associated with an application. This together with `app`
  label should also be used as replicas selector. E.g. `app=kvm-operator,version=1.0.0`. Exceptions can

>   be made to accomodate for adherence to existing selectors upstream.

Reviewer: With `app.giantswarm.io/version` we should update it and state that this is the chart version.

```suggestion
- `version` - value should contain the version of the chart installing the
  application.  Should be applied to every Kubernetes resource in the chart.
  This together with `app` label should also be used as replicas selector. E.g.
  `app=kvm-operator,version=1.0.0`. Exceptions can be made to accommodate for
  adherence to existing selectors upstream.
```

Author: Good catch :+1:

  by the operator to recognize which object it should reconcile. When set on
  nodes it is used to set that information in the status with the
  statusresource.
- `OPERATOR.giantswarm.io/branch` - (informational) branch from which this

>   instance of the `OPERATOR` was built from.

Reviewer: ```suggestion
- `giantswarm.io/branch` - (informational) branch from which this
  code was built from.
```

Author: yeah I know it seems slightly redundant, but just `giantswarm.io/branch` reads weird, i.e. a branch of what

Reviewer: I mean this is what this change in docs is supposed to explain right?

Also it's easier on tooling:

```
kubectl get deploy -n giantswarm -l "app=aws-operator,giantswarm.io/branch=legacy"
````

vs

```
kubectl get deploy -n giantswarm -l "app=aws-operator,aws-operator.giantswarm.io/branch=legacy"
````

For me the first one is more appealing.

Reviewer: Oh yeah you can do:

```
kubectl get deploy -n giantswarm -l "aws-operator.giantswarm.io/branch=legacy"
```

But still I'd like to avoid dynamic label names if not necessary. This is not intuitive IMO.

Reviewer: https://github.com/giantswarm/fmt/pull/100/files#r390885624

Author: Done

- `OPERATOR.giantswarm.io/commit` - (informational) ID of the commit from which

>   this instance of the `OPERATOR` was built from.

Reviewer: ```suggestion
- `giantswarm.io/commit` - (informational) sha of the commit from which
  code was built from.
```

Reviewer: How about `app.giantswarm.io/commit`?

Reviewer: Then we know this is about the app and it isn't dynamic.

Author: Done, thanks!
